Release Godot 4.2 RC 2 and publish release notes (#757)
godotengine/godot-website: collections/_article/release-candidate-godot-4-2-rc-2.md

---
title: "Release candidate: Godot 4.2 RC 2"
excerpt: "The first release candidate brought another crop of regression reports, which we've now fixed!"
categories: ["pre-release"]
author: Rémi Verschelde
image: /storage/blog/covers/release-candidate-godot-4-2-rc-2.webp
image_caption_title: "2D platformer prototype"
image_caption_description: "An open source game by Securas"

> date: 2023-11-24 17:00:00

Author: ```suggestion
date: 2023-11-24 20:00:00
```

---

> Since our first release candidate for Godot 4.2 a week ago, a number of regressions have been reported by new testers who were starting to test using Godot 4.2 in production. So we've been busy this week triaging, testing, debugging, and fixing regressions to reach a point where we're confident that we're ready for the stable release.

Reviewer: ```suggestion
Since our [first release candidate for Godot 4.2](/article/release-candidate-godot-4-2-rc-1/) a week ago, a number of regressions have been reported by new testers who were starting to use Godot 4.2 in production. So we've been busy this week triaging, testing, debugging, and fixing regressions to reach a point where we're confident that we're ready for the stable release.
```
Please double-check I have the correct link syntax! I changed "testers" to avoid having "testers" and "test" in close proximity. Could alternatively say "reported by new testers who were starting to use Godot 4.2 in production"?

Author: I like the latter! Edited your suggestion to commit it.

> Since our first release candidate for Godot 4.2 a week ago, a number of regressions have been reported by new testers who were starting to test using Godot 4.2 in production. So we've been busy this week triaging, testing, debugging, and fixing regressions to reach a point where we're confident that we're ready for the stable release.

Reviewer 2: Missing a link to said RC as per tradition?

> That's where this second release candidate comes in, polished as much as we can and ready to be trialed on all your Godot 4.1 projects, hopefully as a smooth upgrade! If you notice anything off that's not already listed under the ]known issues](#known-issues) below, please make sure to report it quickly so we can keep track of it in our release planning.

Reviewer 3: ```suggestion
That's where this second release candidate comes in, polished as much as we can and ready to be trialed on all your Godot 4.1 projects, hopefully as a smooth upgrade! If you notice anything off that's not already listed under the [known issues](#known-issues) below, please make sure to report it quickly so we can keep track of it in our release planning.
```

This release candidate includes a number of big changes that warrant particular attention:

> - Many users have been running into spurious `slot >= slot_max` errors in their release exports in Godot 4.0 and 4.1, typically around changing scenes or otherwise manipulating nodes in the scene tree. We finally managed to debug and fix it properly ([GH-85280](https://github.com/godotengine/godot/pull/85280)), which should solve a lot of issues, including crashes on Windows. This change should find its way in a 4.1.x maintenance release too.

Reviewer: ```suggestion
- Many users have been running into spurious `slot >= slot_max` errors in their release exports in Godot 4.0 and 4.1, typically around changing scenes or otherwise manipulating nodes in the scene tree. We finally managed to debug and fix this properly ([GH-85280](https://github.com/godotengine/godot/pull/85280)), which should solve a lot of issues, including crashes on Windows. This change should find its way in a 4.1.x maintenance release too.
```

> - For Windows builds, we kept running into issues with MinGW's implementation of `std::thread` and related threading classes. We replaced it with the third-party [mingw-std-threads](https://github.com/meganz/mingw-std-threads) library, which seems to solve the issues we had, and should also improve performance. Integrating such a library so late in the release cycle is a risky change, but we ran out of alternative to workaround those threading issues, and the first test results are promising. Please make sure to test our Windows builds heavily on projects using multithreading.

Reviewer: ```suggestion
- For Windows builds, we kept running into issues with MinGW's implementation of `std::thread` and related threading classes. We replaced it with the third-party [mingw-std-threads](https://github.com/meganz/mingw-std-threads) library, which seems to solve the issues we had, and should also improve performance. Integrating such a library so late in the release cycle is a risky change, but we ran out of alternatives to workaround those threading issues, and the first test results are promising. Please make sure to test our Windows builds heavily on projects using multithreading.
```

> - The tool we made to upgrade 4.1 mesh surfaces to the new 4.2 format works well, but the way it showed itself as an immediate dialog could cause some rendering deadlocks in the editor. So we changed it to a less intrusive warning, which will encourage you to run the tool yourself via the 'Project > Tools > Upgrade Mesh Surfaces..' entry. It's easier to miss, but it's better than a frozen editor.

Reviewer: ```suggestion
- The tool we made to upgrade 4.1 mesh surfaces to the new 4.2 format works well, but the way it showed itself as an immediate dialog could cause some rendering deadlocks in the editor. So we changed it to a less intrusive warning, which will encourage you to run the tool yourself via the 'Project > Tools > Upgrade Mesh Surfaces...' entry. It's easier to miss, but it's better than a frozen editor.
```
I haven't actually checked, but I assume it's three dots rather than two.

> - The tool we made to upgrade 4.1 mesh surfaces to the new 4.2 format works well, but the way it showed itself as an immediate dialog could cause some rendering deadlocks in the editor. So we changed it to a less intrusive warning, which will encourage you to run the tool yourself via the 'Project > Tools > Upgrade Mesh Surfaces..' entry. It's easier to miss, but it's better than having a frozen editor.

Reviewer 3: ```suggestion
- The tool we made to upgrade 4.1 mesh surfaces to the new 4.2 format works well, but the way it showed itself as an immediate dialog could cause some rendering deadlocks in the editor. So we changed it to a less intrusive warning, which will encourage you to run the tool yourself via the 'Project > Tools > Upgrade Mesh Surfaces..' entry. It's easier to miss, but it's better than a frozen editor.
```

> - As a reminder, since beta 5, we provide pre-compiled ARM64 and ARM32 Linux binaries (editor and templates). This RC 2 improves this new set of binaries by temporarily disabling <abbr title="Link Time Optimization">LTO</abbr>, which seemed to cause crashes on at least Raspberry Pi OS. Please try it out if you have such ARM Linux systems!

Reviewer: ```suggestion
- As a reminder, since beta 5, we provide pre-compiled ARM64 and ARM32 Linux binaries (editor and templates). RC 2 improves this new set of binaries by temporarily disabling <abbr title="Link Time Optimization">LTO</abbr>, which seemed to cause crashes on at least Raspberry Pi OS. Please try it out if you have such ARM Linux systems!
```

**13 contributors** submitted **34 improvements** for this release. You can review the complete list of changes with our [interactive changelog](https://godotengine.github.io/godot-interactive-changelog/#4.2-rc2), which contains links to relevant commits and PRs for this and every previous release. Below are the most notable changes compared to 4.2-rc1:

- Animation: Fix seeking bug in AnimationPlayerEditor ([GH-85193](https://github.com/godotengine/godot/pull/85193)).
- Animation: Clear seeked/started flag after seeking/advancing in AnimationPlayer ([GH-85221](https://github.com/godotengine/godot/pull/85221)).
- Animation: Bind `_reset`/`_restore` in AnimationMixer ([GH-85254](https://github.com/godotengine/godot/pull/85254)).
- Animation: Fix TrackCache memory crash ([GH-85266](https://github.com/godotengine/godot/pull/85266)).
- Animation: Perform safe copies in `AnimatedValuesBackup::get_cache_copy()` ([GH-85302](https://github.com/godotengine/godot/pull/85302)).

> - Animation: Fix a crash when trying to restore uncopyable animation tracks ([GH-85308](https://github.com/godotengine/godot/pull/85308)).

Reviewer 2: I would normally add such lines to the line of the original PR without listing it explicitly, since it's a part of the same package.

> - Animation: Fix TrackCache memory crash ([GH-85266](https://github.com/godotengine/godot/pull/85266)).
> - Animation: Perform safe copies in `AnimatedValuesBackup::get_cache_copy()` ([GH-85302](https://github.com/godotengine/godot/pull/85302)).
> - Animation: Fix a crash when trying to restore uncopyable animation tracks ([GH-85308](https://github.com/godotengine/godot/pull/85308)).

Author: ```suggestion
- Animation: Fix TrackCache memory crash ([GH-85266](https://github.com/godotengine/godot/pull/85266), [GH-85302](https://github.com/godotengine/godot/pull/85302), [GH-85308](https://github.com/godotengine/godot/pull/85308)).
```

- Shaders: Make `AMOUNT_RATIO` constant in the shader language specification ([GH-85086](https://github.com/godotengine/godot/pull/85086)).
- Shaders: Set some dialogs in the VisualShader editor to be exclusive ([GH-85205](https://github.com/godotengine/godot/pull/85205)).
- Documentation and translation updates.

This release is built from commit [`1ba920fad`](https://github.com/godotengine/godot/commit/1ba920fada9efc8c4476ded50fe673b8db541366).

## Downloads

{% include articles/download_card.html version="4.2" release="rc2" article=page %}

**Standard build** includes support for GDScript and GDExtension.

**.NET build** (marked as `mono`) includes support for C#, as well as GDScript and GDExtension.
- .NET build requires [.NET SDK 6.0](https://dotnet.microsoft.com/en-us/download/dotnet/6.0), [7.0](https://dotnet.microsoft.com/en-us/download/dotnet/7.0), or [8.0](https://dotnet.microsoft.com/en-us/download/dotnet/8.0) installed in a standard location.
- To export to Android, .NET 7.0 or later is required. To export to iOS, .NET 8.0 is required. Make sure to set the target framework in the `.csproj` file.

{% include articles/prerelease_notice.html %}

## Known issues

> We are aware of some regressions which still affect this release candidate, but which we may not be able to solve in time for the stable release. They're still tracked as high priority and should be fixed soon after the release with backports to 4.2.x maintenance releases.

Reviewer: ```suggestion
We are aware of some regressions which still affect this release candidate, but which we may not be able to solve in time for the stable release. They're still tracked as high priority and should be fixed soon in 4.2.x maintenance releases.
```
Although I know the development methodology (where you push to master and backport relevant things to older branches), I think it's clearer as I've suggested above.

- TileMap editing performance on macOS got significantly worse when using the ANGLE backend for the Compatibility renderer ([GH-84591](https://github.com/godotengine/godot/issues/84591)). We have good hunches for what causes the issue but it will take some time to solve properly. In the meantime, you can change the `rendering/gl_compatibility/driver.macos` project setting to `opengl3` instead of the new default `opengl3_angle` to restore the 4.1 behavior if you are affected by this regression.

> - Capture track in animations no longer work after the unification of the AnimationPlayer and AnimationTree processing in AnimationMixer ([GH-83166](https://github.com/godotengine/godot/issues/83166)). This feature was specific to AnimationPlayer and was lost in the refactoring. We're evaluating what's needed to restore this feature.

Reviewer: I'm not sure if this should be "Capture tracks in animations no longer work" or "Capture track in animations no longer works", maybe?

Author: The former makes sense.

> - Capture track in animations no longer work after the unification of the AnimationPlayer and AnimationTree processing in AnimationMixer ([GH-83166](https://github.com/godotengine/godot/issues/83166)). This feature was specific to AnimationPlayer and was lost in the refactoring. We're evaluating what's needed to restore this feature.

Author: ```suggestion
- Capture tracks in animations no longer work after the unification of the AnimationPlayer and AnimationTree processing in AnimationMixer ([GH-83166](https://github.com/godotengine/godot/issues/83166)). This feature was specific to AnimationPlayer and was lost in the refactoring. We're evaluating what's needed to restore this feature.
```
